### ISSUE TITLE
The solution to the slow building problem

### ISSUE THREAD

**Issue by smachave** [Jimdo/typings-for-css-modules-loader#48]
At least in my opinion,
```js
plugins: [
    new webpack.WatchIgnorePlugin([
      /css\.d\.ts$/
    ]),
    ...
  ]
```
isn't the correct way to resolve this problem. In my case when I appended that line, it started to ignore `*.css.d.ts` files completely (sounds obvious, but read on). Which meant that, once `typings-for-css-modules-loader` created the typings definition file, **Webpack didn't recompile**! So whenever I added a new class in the css file, I got an error. It didn't occur if I changed an attribute inside a preexisting class since the typings definition for it already existed.

The correct way would be to not make Webpack ignore those files at all! What you could do is, run a sequential task and generate the typings defs before starting webpack using something like [this](https://github.com/Quramy/typed-css-modules).

But even then you do get an error once a new class has been added. The pass that generates the error is also responsible for creating the typings def file, and when that happens, webpack realizes that a change has been made to one of the files and recompiles, but this time it's a success! 😁

**Comment by zhenwenc** [comment]
Hi @tkshnwesper , I am wondering without the `WatchIgnorePlugin`, do you experience multiple re-compiling of the typescript files right after starting webpack dev server? As webpack loads all `.css` files, `typings-for-css-modules-loader` creates the typings definition files (even they are already exists), which triggers webpack to rebuild...

**Comment by smachave** [comment]
Yes, multiple recompilations do take place. It's not a bad trade off in my opinion, since all I want it to do is work.

**Comment by zhenwenc** [comment]
Thanks for the reply. My project have many css files, it triggers quite a lot of recompilations, which makes me worried.

**Comment by smachave** [comment]
What I do now is watch for changes to my CSS files and then compile them to type defs using typed-css-modules exclusive of webpack. Although i wouldn't recommend it, it still does the job :)

**Comment by reiv** [comment]
I took a crack at solving this one. Basically, the issue for me was that `ts-loader` keeps a cache of source files, including the generated `.d.ts` declarations.  It does this by hooking into webpack's file monitoring system, which is also used by webpack internally for triggering compilation in watch mode.  Now when adding `WatchIgnorePlugin` to the mix, the timestamps of ignored files are simply set to `1`, meaning they do not trigger compilation; however, this also means that `ts-loader` does not pick them up when (re-)generated by `typings-for-css-modules-loader` -- it just uses the cached version (i.e.  the one from the initial build). My (hacky) workaround is to add a plugin which simply manually re-adds those timestamps so that `ts-loader` picks them up again.  

Edit: @developit wrote an [updated version](https://github.com/Jimdo/typings-for-css-modules-loader/issues/48#issuecomment-372378406) that also works with Webpack 4; use that instead!

In `webpack.config.js`:

```javascript
var fs = require('fs');
var webpack = require('webpack');

function WatchTimestampsPlugin(patterns) {
  this.patterns = patterns;
};

WatchTimestampsPlugin.prototype.apply = function(compiler) {
  compiler.plugin('watch-run', function(watch, callback) {
    var patterns = this.patterns;
    Object.keys(watch.compiler.fileTimestamps).forEach(function (filepath) {
      if (patterns.some(function (pat) {
        return pat instanceof RegExp ? pat.test(filepath) : filepath.indexOf(pat) === 0;
      })) {
        watch.compiler.fileTimestamps[filepath] = fs.statSync(filepath).mtime;
      }
    });
    callback();
  }.bind(this));
}

module.exports = {
  entry: './src/index.ts',
  output: {
    filename: 'bundle.js'
  },
  module: {
    loaders: [
      {
        test: /\.tsx?$/,
        loaders: ['ts-loader'],
        exclude: /node_modules/
      },
      {
        test: /\.css$/,
        loader: 'typings-for-css-modules-loader?modules&namedExport&camelCase'
      }
    ]
  },
  resolve: {
    extensions: ['.js', '.ts', '.tsx'],
  },
  plugins: [
    // Ignore declaration files generated by typings-for-css-module-loader.
    new webpack.WatchIgnorePlugin([
      /css\.d\.ts$/,
    ]),
    // ...but ensure they are still picked up by ts-loader's caching mechanism.
    new WatchTimestampsPlugin([
      /css\.d\.ts$/,
    ])
  ],
};
```

**Comment by zhenwenc** [comment]
Thanks @reiv ! Your plugin nicely solve my issue. :-)

**Comment by developit** [comment]
Just an update for those coming across this - as of Webpack 4, `watch.compiler.fileTimestamps` is now `watch.fileTimestamps`, and is a `Map` instead of a POJO.  Here's an updated plugin:

```js
let fs = require('fs');
module.exports = WatchTimestampsPlugin;
function WatchTimestampsPlugin(patterns) {
	this.patterns = patterns;
}

WatchTimestampsPlugin.prototype.apply = function (compiler) {
	compiler.plugin('watch-run', (watch, callback) => {
		let patterns = this.patterns;
		let timestamps = watch.fileTimestamps || watch.compiler.fileTimestamps;

		Object.keys(timestamps).forEach( filepath => {
			if (patterns.some( pat => pat instanceof RegExp ? pat.test(filepath) : !~filepath.indexOf(pat) )) {
				let time = fs.statSync(filepath).mtime;
				if (timestamps instanceof Map) timestamps.set(filepath, time);
				else timestamps[filepath] = time;
			}
		});
		callback();
	});
};
```
